internal/app/appauth/storage.go

```diff
@@ -0,0 +1,347 @@
+	}
+
+	request := &AuthRequest{
+		ID:           ulid.MustNew(ulid.Timestamp(ti), entropy).String(),
+		ClientID:     authReq.ClientID,
+		subject:      "",
+		code:         "", // Will be set after /authorize/callback success
+		state:        authReq.State,
+		scopes:       authReq.Scopes,
+		audiences:    audiences,
+		ResponseType: authReq.ResponseType,
+		Nonce:        authReq.Nonce,
+		RedirectURI:  authReq.RedirectURI,
+		createdAt:    time.Now().UTC(),
+		authorizedAt: nil,
+	}
+	if authReq.CodeChallenge != "" {
+		request.CodeChallenge = &oidc.CodeChallenge{
+			Challenge: authReq.CodeChallenge,
+			Method:    authReq.CodeChallengeMethod,
+		}
+	}
+
+	s.requests[request.ID] = *request
+	return request, nil
+}
+
+func (s *Storage) AuthRequestByID(_ context.Context, requestID string) (op.AuthRequest, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if requestID == "" {
+		return nil, errors.New("invalid id")
+	}
+	request, exists := s.requests[requestID]
+	if !exists {
+		return nil, errors.New("not found")
+	}
+	return &request, nil
+}
+
+func (s *Storage) AuthRequestByCode(_ context.Context, code string) (op.AuthRequest, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if code == "" {
+		return nil, errors.New("invalid code")
+	}
+	for _, request := range s.requests {
+		if request.GetCode() == code {
+			return &request, nil
+		}
+	}
+	return nil, errors.New("invalid code")
+}
+
+func (s *Storage) AuthRequestBySubject(_ context.Context, subject string) (op.AuthRequest, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if subject == "" {
+		return nil, errors.New("invalid subject")
+	}
+	for _, request := range s.requests {
+		if request.GetSubject() == subject {
+			return &request, nil
+		}
+	}
+	return nil, errors.New("invalid subject")
+}
+
+func (s *Storage) SaveAuthCode(ctx context.Context, requestID, code string) error {
+
+	request, err := s.AuthRequestByID(ctx, requestID)
+	if err != nil {
+		return err
+	}
+	request2 := request.(*AuthRequest)
+	request2.code = code
+	err = s.updateRequest(ctx, requestID, *request2)
+	return err
+}
+
+func (s *Storage) DeleteAuthRequest(_ context.Context, requestID string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	delete(s.clients, requestID)
+	return nil
+}
+
+func (s *Storage) CreateAccessToken(_ context.Context, _ op.TokenRequest) (string, time.Time, error) {
+	return "id", time.Now().UTC().Add(5 * time.Hour), nil
+}
+
+func (s *Storage) CreateAccessAndRefreshTokens(_ context.Context, request op.TokenRequest, _ string) (accessTokenID string, newRefreshToken string, expiration time.Time, err error) {
+	authReq := request.(*AuthRequest)
+	return "id", authReq.ID, time.Now().UTC().Add(5 * time.Minute), nil
+}
+
+func (s *Storage) TokenRequestByRefreshToken(ctx context.Context, refreshToken string) (op.RefreshTokenRequest, error) {
+	r, err := s.AuthRequestByID(ctx, refreshToken)
+	return r.(op.RefreshTokenRequest), err
```

Comment:
```suggestion
	r, err := s.AuthRequestByID(ctx, refreshToken)
	if err != nil {
		return nil, err
	}
	return r.(op.RefreshTokenRequest), nil
```

Note that `r.(xxx)` can occur panic, so this code means r must be always `op.RefreshTokenRequest`